src/components/GeneralProductCard/GeneralProductCard.tsx

```diff
@@ -1,33 +1,85 @@
 'use client';
 
+import { useCallback, useState } from 'react';
 import Image from 'next/image';
+import Checkmark from '~/../public/authenticated/checkmark.svg';
 import NotFavourite from '~/../public/authenticated/not-favourite.svg';
 import Favourite from '~/../public/authenticated/favourite.svg';
 import Button from '~/components/Button';
+import { useApi } from '~/hooks/useApi';
+import { useAuthLayout } from '~/hooks/useAuthLayout';
+import Modal from 'react-modal';
+import closeModal from '~/../public/authenticated/close-modal.svg';
 
 export default function GeneralProductCard({
+  id,
   image,
   price,
   name,
   state,
   favourite,
   firstProduct = false,
   lastProduct = false,
+  cart,
+  setCart,
 }: {
+  id: string;
   image: string;
   price: number;
   name: string;
   state: string;
   favourite: boolean;
   firstProduct?: boolean;
   lastProduct?: boolean;
+  cart: Array<{ quantity: number; product: { id: string } }>;
+  setCart: Function;
 }) {
+  const { session } = useAuthLayout();
+  const { doPost, doGet } = useApi({ session });
+  const [isOpen, setIsOpen] = useState(false);
+
+  const modalStyles = {
+    content: {
+      margin: 'auto',
+      width: 'fit-content',
+      height: 'fit-content',
+      top: 32,
+      padding: 0,
+      display: 'inline-block',
+    },
+  };
+
   const stateBackgroundColour =
     state === 'Restored' ? 'bg-[#559F21]' : 'bg-[#F2C94C]';
   const favouriteIcon = favourite ? Favourite : NotFavourite;
 
   let extraClasses = '';
 
+  const quantityForProduct = useCallback(
+    (id: string): number => {
+      const item = cart.find((cartItem) => {
+        return cartItem.product?.id === id;
+      });
+
+      if (!item) {
+        return 0;
+      } else {
+        return item.quantity;
+      }
+    },
+    [cart]
+  );
+
+  const addProductToCart = async () => {
+    await doPost({
+      endpoint: '/api/shopping-cart/products/',
+      body: { product: id, quantity: quantityForProduct(id) + 1 },
+    });
+    const newCart = await doGet({ endpoint: '/api/shopping-cart/' });
+    setCart(newCart['order_products']);
+    setIsOpen(true);
+  };
+
```

Comment:
Regarding the logic to add remove items from the cart, this should remain in the parent component since there is where the state is defined and passing down the cart state to each product in the cart list may trigger unnecessary rerenders.

The refactor for that would be to trigger a callback when the product is added (instead of onChangeCart it would be something like onProductAdded) and this will trigger the logic in the parent component to add the product.

For this logic you need to identify which is the producted that was clicked so when the callback was triggered you can check which is the index of the product in the parent component and add the correct product to the cart. This could also be done using the product id as well as sending the product object as argument from the child component. 

Let me know if you want we can have a call to explain this better